dali/test/python/test_RN50_external_source_parallel_utils.py

```diff
@@ -0,0 +1,244 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import argparse
+import time
+import cv2
+import numpy as np
+
+import nvidia.dali as dali
+import nvidia.dali.types as types
+from nvidia.dali.plugin.base_iterator import LastBatchPolicy
+
+from test_utils import AverageMeter
+
+
+class FilesDataSet:
+
+    def __init__(self, data_path):
+        files = self.get_files_list(data_path)
+        self.data_path = data_path
+        self.classes_no = len(files)
+        self.files_map = {}
+
+        counter = 0
+        for class_no, (_, samples) in enumerate(files):
+            for sample in samples:
+                self.files_map[counter] = (sample, class_no)
+                counter += 1
+        self.size = counter
+
+    @classmethod
+    def get_files_list(cls, data_path):
+        dirs = [
+            (dir_name, [
+                file_path for file_path
+                in map(lambda name: os.path.join(dir_path, name), os.listdir(dir_path))
+                if os.path.isfile(file_path)
+            ])
+            for (dir_name, dir_path)
+            in map(lambda name: (name, os.path.join(data_path, name)), os.listdir(data_path))
+            if os.path.isdir(dir_path)
+        ]
+        dirs.sort(key=lambda dir_files: dir_files[0])
+        for _, files in dirs:
+            files.sort()
+        return dirs
+
+    def __len__(self):
+        return self.size
+
+    def get_sample(self, sample_idx, epoch_idx):
+        if sample_idx >= self.size:
+            raise StopIteration
+        return self.files_map[sample_idx]
+
+
+class ShuffledFilesDataSet(FilesDataSet):
+
+    def __init__(self, data_path):
+        super().__init__(data_path)
+        self.rng = np.random.default_rng(seed=42)
+        self.epoch_idx = 0
+        self.perm = self.rng.permutation(self.size)
+
+    def get_sample(self, sample_idx, epoch_idx):
+        if self.epoch_idx != epoch_idx:
+            self.perm = self.rng.permutation(self.size)
+            self.epoch_idx = epoch_idx
+        if sample_idx >= self.size:
+            raise StopIteration
+        return super().get_sample(self.perm[sample_idx], epoch_idx)
+
+
+class SampleLoader(object):
+
+    DATA_SET = FilesDataSet
+
+    def __init__(self, data_path):
+        self.data_path = data_path
+        self.data_set = self.create_data_set()
+
+    def __getattr__(self, key):
+        if key == 'data_set':
+            self._data_set = self._data_set or self.create_data_set()
+            self.__dict__['data_set'] = self._data_set
+            return self._data_set
+        raise AttributeError
+
+    def create_data_set(self):
+        return self.DATA_SET(data_path=self.data_path)
+
+    def read_file(self, file_path):
+        with open(file_path, 'rb') as f:
+            return f.read()
+
+    def __call__(self, sample_info):
+        file_path, class_no = self.data_set.get_sample(sample_info.idx_in_epoch, sample_info.iteration)
```

Comment:
I'm not sure if `sample_info.iteration` should map to `epoch_idx`.

Reply:
I think it should be up to the loader to know how many samples it has provided already and when to move to the next epoch (rise StopIteration).